src/cargo/ops/cargo_package.rs

```diff
@@ -309,62 +312,125 @@ fn build_ar_list(
-                    "{} `{}` appears to be a path outside of the package, \
-                            but there is already a file named `{}` in the root of the package. \
-                            The archived crate will contain the copy in the root of the package. \
-                            Update the {} to point to the path relative \
-                            to the root of the package to remove this warning.",
-                    label,
-                    file_path.display(),
-                    file_name.to_str().unwrap(),
-                    label,
-                ))?;
-            } else {
-                result.push(ArchiveFile {
-                    rel_path: PathBuf::from(file_name),
-                    rel_str: file_name.to_str().unwrap().to_string(),
-                    contents: FileContents::OnDisk(abs_file_path),
-                })
+    } else {
+        // If the file exists somewhere outside of the package we will need
+        // to add it to the root
+        //
+        // Calculate if there is name collision with the outside package
+        // (warning) or if there is a exact path collision (error)
+        let file_name = abs_file_path.file_name().unwrap();
+        let mut name_collision = false;
+        let mut exact_path_collision = false;
+        for ar in files.iter() {
+            if ar.rel_path.file_name().unwrap() == file_name {
+                name_collision = true;
+                if ar.rel_path == pkg.root() {
+                    exact_path_collision = true;
+                    break;
+                }
             }
         }
+
+        if name_collision {
+            // Warn
+            ws.config().shell().warn(&format!(
+                "{} `{}` appears to be a path outside of the package, \
+                    but there is already a file named `{}` in the package. \
+                    The archived crate will contain the copy in the root of the package. \
+                    Update the {} to point to a path relative \
+                    to the root of the package to remove this warning.",
+                label,
+                file_path.display(),
+                file_name.to_str().unwrap(),
+                label,
+            ))?;
+        } else if exact_path_collision {
+            // Error (panic)
+            ws.config().shell().error(&format!(
+                "{} `{}` appears to be a path outside of the package, \
+                    but there is already a file `{}` in the root of the package. \
+                    because the default behaviour for external files referenced by \
+                    the pacakge is to put them on the root and there is already a file \
+                    with that name on the package this is a fatal error. ",
+                label,
+                file_path.display(),
+                file_name.to_str().unwrap(),
+            ))?;
+            std::process::exit(-1); // Don't know if the code matters
```

Comment:
Is this the proper way to exit when you log an error with `ws.config().shell().error(...)` or is there a more refined way?